### PR TITLE
WIP: Detect too old WebViews

### PR DESCRIPTION
**Pull request by strseb**
This is an idea i came up to handle #344. 
Even though i think it's unlikely that a user on a real device that has the play store will encounter this, a fallback is always good.
So what i try to do here is to detect old webviews and then use the DesktopAuth. 
In that case we would open the users default browser instead of a webview, which hopefully is way newer. 
Currently the user would need to navigate back to the app by hand, but still this is way better then not beeing able to login in at all.

### CODE REVIEW

File: src/platforms/android/androidauthenticationlistener.cpp [hunk 1]
```diff
@@ -8,15 +8,19 @@
 #include "mozillavpn.h"
 #include "platforms/android/androidutils.h"
 
+#include <QAndroidJniObject>
+#include <QtAndroid>
+#include <jni.h>
+#include "tasks/authenticate/desktopauthenticationlistener.h"
```
Comment (bakulf) [inline src/platforms/android/androidauthenticationlistener.cpp:14]: move this to the rest of the local headers

File: src/platforms/android/androidauthenticationlistener.cpp [hunk 2]
```diff
@@ -31,5 +35,21 @@ void AndroidAuthenticationListener::start(MozillaVPN* vpn, QUrl& url,
...
-  AndroidUtils::instance()->startAuthentication(this, url);
+  QAndroidJniObject activity = QtAndroid::androidActivity();
+  jboolean supported = QAndroidJniObject::callStaticMethod<jboolean>(
+      "com/mozilla/vpn/PackageManagerHelper", "isWebViewSupported",
+      "(Landroid/content/Context;)Z", activity.object());
+  if (supported) {
+    AndroidUtils::instance()->startAuthentication(this, url);
+    return;
+  }
+  m_legacyAuth = new DesktopAuthenticationListener(this);
```
Comment (bakulf) [inline src/platforms/android/androidauthenticationlistener.cpp:46]: Well, this doesn't fix the issue, right? We are just moving the problem elsewhere.
Of course, if the user has another default browser, the operation _could_ work.

File: src/platforms/android/androidauthenticationlistener.cpp [hunk 3]
```diff
@@ -31,5 +35,21 @@ void AndroidAuthenticationListener::start(MozillaVPN* vpn, QUrl& url,
...
+  m_legacyAuth->start(vpn, url, query);
+
+  connect(m_legacyAuth, &AuthenticationListener::completed,
```
Comment (bakulf) [inline src/platforms/android/androidauthenticationlistener.cpp:49]:
```
connect(m_legacyAuth, &AuthenticationListener::completed, this, &AndroidAuthenticationListener::completed);
connect(m_legacyAuth, &AuthenticationListener::failed, this, &AndroidAuthenticationListener::failed);
```

File: src/platforms/android/androidauthenticationlistener.cpp (hunk 2 above)
Comment (bakulf) [inline src/platforms/android/androidauthenticationlistener.cpp:46]: No need to store m_legacyAuth in the object. Just do:

```
DesktopAuhenticationListener* legacyAuth = new DesktopAuthenticationListener(this);
connect...

File: src/platforms/android/androidauthenticationlistener.h [hunk 4]
```diff
@@ -17,6 +18,9 @@ class AndroidAuthenticationListener final : public AuthenticationListener {
   ~AndroidAuthenticationListener();
 
   void start(MozillaVPN* vpn, QUrl& url, QUrlQuery& query) override;
+
+ private:
+  DesktopAuthenticationListener* m_legacyAuth;
```
Comment (bakulf) [inline src/platforms/android/androidauthenticationlistener.h:23]: Remove this

File: src/platforms/android/androidauthenticationlistener.h [hunk 5]
```diff
@@ -6,6 +6,7 @@
 #define ANDROIDAUTHENTICATIONLISTENER_H
 
 #include "authenticationlistener.h"
+#include "tasks/authenticate/desktopauthenticationlistener.h"
```
Comment (bakulf) [inline src/platforms/android/androidauthenticationlistener.h:9]: Remove this.

**Review by bakulf — CHANGES_REQUESTED** [review]

File: src/platforms/android/androidauthenticationlistener.cpp (hunk 2 above)
Comment (strseb) [reply src/platforms/android/androidauthenticationlistener.cpp:46]: Indeed, if the default browser is too old this wont work. But the user (and support) can help themselves here as they can just switch default browsers.
- Note the old client does not have any checks of this sort, it's just uses the installed firefox( if available ) or google chrome.

File: src/platforms/android/androidauthenticationlistener.cpp [hunk 6]
```diff
@@ -15,8 +20,7 @@ Logger logger(LOG_ANDROID, "AndroidAuthenticationListener");
 AndroidAuthenticationListener::AndroidAuthenticationListener(QObject* parent)
     : AuthenticationListener(parent) {
   MVPN_COUNT_CTOR(AndroidAuthenticationListener);
-
-  logger.log() << "Android authentication listener";
+  logger.log() << "Android authentication litener";
```
Comment (bakulf) [inline src/platforms/android/androidauthenticationlistener.cpp:23]: typo

**Review by bakulf — APPROVED** [review]

**Pull request merged**